### PR TITLE
[Service Bus Client] Azure.Core Shared Access Credentials

### PR DESCRIPTION
**Pull request by jsquire**
# Summary

The focus of these changes is to support the `AzureNamedKeyCredential` and `AzureSasCredential` types for client authorization, replacing the unreleased Service Bus-specific equivalent.

# Last Upstream Rebase

Wednesday, March 24, 1pm (EDT)

### CODE REVIEW

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Authorization/ServiceBusTokenCredential.cs [hunk 1]
```diff
@@ -15,47 +15,33 @@ namespace Azure.Messaging.ServiceBus.Authorization
     ///
     internal class ServiceBusTokenCredential : TokenCredential
     {
-        /// <summary>
-        ///   The Service Bus resource to which the token is intended to serve as authorization.
-        /// </summary>
-        ///
-        public string Resource { get; }
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Authorization/ServiceBusTokenCredential.cs:22]: I don't recall what the original plan was, but this was never referenced.

File: sdk/servicebus/Azure.Messaging.ServiceBus/Azure.Messaging.ServiceBus.sln [hunk 2]
```diff
@@ -31,8 +31,6 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "samples", "samples", "{8B8C
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Azure.Core.Experimental", "..\..\core\Azure.Core.Experimental\src\Azure.Core.Experimental.csproj", "{5871D9C6-F2DF-4F05-B29A-6C0D8709784A}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Azure.Core", "..\..\core\Azure.Core\src\Azure.Core.csproj", "{9CD1905A-19B9-4F71-8BE6-D25CF49DA5B3}"
-EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Azure.Core.Amqp", "..\..\core\Azure.Core.Amqp\src\Azure.Core.Amqp.csproj", "{2ADA26CA-77E5-4793-927A-A6185FD8AA29}"
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/Azure.Messaging.ServiceBus.sln:34]: Why remove?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Administration/ServiceBusAdministrationClient.cs [hunk 3]
```diff
@@ -126,31 +124,42 @@ public ServiceBusAdministrationClient(string connectionString)
...
 
-            HttpPipeline pipeline = HttpPipelineBuilder.Build(options);
-            _clientDiagnostics = new ClientDiagnostics(options);
+        /// <summary>
+        /// Initializes a new <see cref="ServiceBusAdministrationClient"/> which can be used to perform administration operations on ServiceBus entities.
+        /// </summary>
+        ///
+        /// <param name="fullyQualifiedNamespace">The fully qualified Service Bus namespace to connect to.  This is likely to be similar to <c>{yournamespace}.servicebus.windows.net</c>.</param>
+        /// <param name="credential">The <see cref="AzureSasCredential"/> to use for authorization.  Access controls may be specified by the Service Bus namespace or the requested Service Bus entity, depending on Azure configuration.</param>
+        internal ServiceBusAdministrationClient(
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Administration/ServiceBusAdministrationClient.cs:143]: Isn't this meant to be public?

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/CorrelationFilterPropertiesAsync.json [hunk 4]
```diff
@@ -1,16 +1,16 @@
 {
   "Entries": [
     {
-      "RequestUri": "https://jolovservicebus.servicebus.windows.net/006bafbe?api-version=2017-04",
+      "RequestUri": "https://omqsquire.servicebus.windows.net/006bafbe?api-version=2017-04",
       "RequestMethod": "PUT",
       "RequestHeaders": {
         "Authorization": "Sanitized",
         "Content-Length": "843",
         "Content-Type": "application/atom\u002Bxml",
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/CorrelationFilterPropertiesAsync.json:9]: What is the reasoning behind re-recording all of these? I would expect we may have new live tests but existing ones should pass playback without re-recording.

File: sdk/servicebus/Azure.Messaging.ServiceBus/Azure.Messaging.ServiceBus.sln (hunk 2 above)
Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/Azure.Messaging.ServiceBus.sln:34]: We should be using a package reference.   What would the need be for having it as part of the solution?

**Comment by JoshLove-msft** [comment]
There don't seem to be API changes here.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Administration/ServiceBusAdministrationClient.cs (hunk 3 above)
Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Administration/ServiceBusAdministrationClient.cs:143]: Yup.... whoops.   I'll fix that up once the live suite completes.

File: sdk/servicebus/Azure.Messaging.ServiceBus/Azure.Messaging.ServiceBus.sln (hunk 2 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/Azure.Messaging.ServiceBus.sln:34]: It is just a convenience. Most solutions include it.

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/CorrelationFilterPropertiesAsync.json (hunk 4 above)
Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/CorrelationFilterPropertiesAsync.json:9]: I haven't got a good grasp on the recording framework.   I had fails.  I set the mode to Record.  I added whatever got spit out.    No objection to changing, but how do?

Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/CorrelationFilterPropertiesAsync.json:9]: Can you include what the failures were?

**Comment by jsquire** [comment]
> There don't seem to be API changes here.

The public surface changed (see `ServiceBusClient` and `ServiceBusManagementClient`) but API Gen is spitting out the same old version of things and Analyze doesn't seem unhappy.  I'm not sure if it has to do with there being a beta out there already and we're riffing on the beta.    I don't know.

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/CorrelationFilterPropertiesAsync.json (hunk 4 above)
Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/CorrelationFilterPropertiesAsync.json:9]: Not sure that I understand what you're looking for, but I was getting flagged for not having recordings for the Sas-based tests.

File: sdk/servicebus/Azure.Messaging.ServiceBus/Azure.Messaging.ServiceBus.sln (hunk 2 above)
Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/Azure.Messaging.ServiceBus.sln:34]: Your call;  I don't feel strongly, just cleaning up after I removed the unneeded project reference.

**Comment by jsquire** [comment]
> > There don't seem to be API changes here.
> 
> The public surface changed (see `ServiceBusClient` and `ServiceBusManagementClient`) but API Gen is spitting out the same old version of things and Analyze doesn't seem unhappy. I'm not sure if it has to do with there being a beta out there already and we're riffing on the beta. I don't know.

Or, it's because I forgot to flip the named key constructors to `public` and copied the SAS constructors from them, so they're `internal` too.   /sigh

File: sdk/servicebus/Azure.Messaging.ServiceBus/Azure.Messaging.ServiceBus.sln [hunk 5]
```diff
@@ -31,12 +31,12 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "samples", "samples", "{8B8C
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Azure.Core.Experimental", "..\..\core\Azure.Core.Experimental\src\Azure.Core.Experimental.csproj", "{5871D9C6-F2DF-4F05-B29A-6C0D8709784A}"
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/Azure.Messaging.ServiceBus.sln:32]: Re-adding `Azure.Core` changed the ordering and GUIDs.   I'll push a revert to `master` once the test suite completes.

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/AuthenticateWithSharedKeyCredentialAsync.json [hunk 6]
```diff
@@ -7,10 +7,10 @@
         "Authorization": "Sanitized",
         "Content-Length": "851",
         "Content-Type": "application/atom\u002Bxml",
-        "traceparent": "00-716054ab44833347a8ea439361d29ade-e25caad0d4c69345-00",
+        "traceparent": "00-6e9fe8fdd6c95046a50542def9d45550-01ea95b75dc2d64a-00",
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/AuthenticateWithSharedKeyCredentialAsync.json:10]: This one still seems to have been re-recorded.

Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/AuthenticateWithSharedKeyCredentialAsync.json:10]: That was intentional due to the credential type change, but now that you mention it, the changes up the stack probably don't need to be reflected down this way.   Let me try reverting and see if it causes a failure.

Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/AuthenticateWithSharedKeyCredentialAsync.json:10]: Looks like I don't need it.   I'll push a revert once the test runs finish.

File: sdk/servicebus/Azure.Messaging.ServiceBus/api/Azure.Messaging.ServiceBus.netstandard2.0.cs [hunk 7]
```diff
@@ -59,6 +59,10 @@ public partial class ServiceBusClient : System.IAsyncDisposable
     {
         protected ServiceBusClient() { }
         public ServiceBusClient(string connectionString) { }
+        public ServiceBusClient(string fullyQualifiedNamespace, Azure.AzureNamedKeyCredential credential) { }
+        public ServiceBusClient(string fullyQualifiedNamespace, Azure.AzureNamedKeyCredential credential, Azure.Messaging.ServiceBus.ServiceBusClientOptions options) { }
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/api/Azure.Messaging.ServiceBus.netstandard2.0.cs:63]: Should we collapse these into a single ctor as it is probably a more advanced scenario?

File: sdk/servicebus/Azure.Messaging.ServiceBus/api/Azure.Messaging.ServiceBus.netstandard2.0.cs [hunk 8]
```diff
@@ -59,6 +59,10 @@ public partial class ServiceBusClient : System.IAsyncDisposable
...
+        public ServiceBusClient(string fullyQualifiedNamespace, Azure.AzureSasCredential credential) { }
+        public ServiceBusClient(string fullyQualifiedNamespace, Azure.AzureSasCredential credential, Azure.Messaging.ServiceBus.ServiceBusClientOptions options) { }
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/api/Azure.Messaging.ServiceBus.netstandard2.0.cs:65]: Same comment here - I think we want to make options a default parameter.

File: sdk/servicebus/Azure.Messaging.ServiceBus/api/Azure.Messaging.ServiceBus.netstandard2.0.cs [hunk 9]
```diff
@@ -712,6 +716,10 @@ public partial class ServiceBusAdministrationClient
     {
         protected ServiceBusAdministrationClient() { }
         public ServiceBusAdministrationClient(string connectionString) { }
+        public ServiceBusAdministrationClient(string fullyQualifiedNamespace, Azure.AzureNamedKeyCredential credential) { }
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/api/Azure.Messaging.ServiceBus.netstandard2.0.cs:719]: And same comment here.

File: sdk/servicebus/Azure.Messaging.ServiceBus/api/Azure.Messaging.ServiceBus.netstandard2.0.cs (hunk 8 above)
Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/api/Azure.Messaging.ServiceBus.netstandard2.0.cs:65]: I'm good with that, and it's what we do elsewhere;  I didn't change the pattern that was already in place.     If we're doing these, I think that we want to do the same for the `TokenCredential` overload as well, no?

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/Client/ServiceBusClientLiveTests.cs [hunk 10]
```diff
@@ -84,13 +84,13 @@ await using (var client = new ServiceBusClient(TestEnvironment.FullyQualifiedNam
...
         [Test]
-        public async Task ClientCanConnectUsingSharedKeyCredentialWithSharedKey()
+        public async Task ClientCanConnectUsingSharedKeyCredential()
         {
             await using (var scope = await ServiceBusScope.CreateWithQueue(enablePartitioning: true, enableSession: false))
             {
                 var options = new ServiceBusClientOptions();
                 var audience = ServiceBusConnection.BuildConnectionResource(options.TransportType, TestEnvironment.FullyQualifiedNamespace, scope.QueueName);
-                var credential = new ServiceBusSharedAccessKeyCredential(TestEnvironment.SharedAccessKeyName, TestEnvironment.SharedAccessKey);
+                var credential = new AzureNamedKeyCredential(TestEnvironment.SharedAccessKeyName, TestEnvironment.SharedAccessKey);
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/tests/Client/ServiceBusClientLiveTests.cs:93]: Would it be possible to have a live test where we actually update the key after the client is already constructed and after we've already done some operations with the initial key? Maybe we could just update it to use the secondary key from the namespace.

Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/Client/ServiceBusClientLiveTests.cs:93]: We validate this in the `SharedAccessTokenCredentialTests`.   I don't see enough value in doing this as a Live to run the additional infrastructure through the test resources to query and inject a second key.

**Review by JoshLove-msft — APPROVED** [review]

**Pull request merged**